homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,267 @@
+                                         radius=radius))
+        elif _filter['filter'] == FILTER_NAME_LOWPASS:
+            time_constant = _filter.get(CONF_FILTER_TIME_CONSTANT,
+                                        DEFAULT_FILTER_TIME_CONSTANT)
+            filters.append(LowPassFilter(window_size=window_size,
+                                         precision=precision,
+                                         time_constant=time_constant))
+
+    sensors.append(SensorFilter(hass, name, entity_id, filters))
+
+    async_add_devices(sensors)
+
+
+class SensorFilter(Entity):
+    """Representation of a Filter Sensor."""
+
+    filters = [FILTER_NAME_LOWPASS, FILTER_NAME_OUTLIER]
+
+    def __init__(self, hass, name, entity_id, filters):
+        """Initialize the sensor."""
+        self.hass = hass
+        self._name = name
+        self._entity = entity_id
+        self._unit_of_measurement = None
+        self._state = None
+        self._filters = filters
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def filter_sensor_state_listener(entity, old_state, new_state):
+            """Handle device state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+
+            self._state = new_state.state
+            for filt in self._filters:
+                try:
+                    filtered_state = filt.filter_state(self._state)
+                    _LOGGER.debug("%s(%s) -> %s", filt.name, self._state,
+                                  filtered_state)
+                    self._state = filtered_state
+                    filt.states.append(filtered_state)
+                except ValueError:
+                    _LOGGER.warning("Could not convert state to number")
+
+            self.async_schedule_update_ha_state(True)
+
+        async_track_state_change(
+            self.hass, self._entity, filter_sensor_state_listener)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return ICON
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit_of_measurement of the device."""
+        return self._unit_of_measurement
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        state_attr = {
+            ATTR_ENTITY_ID: self._entity
+        }
+        for filt in self._filters:
+            state_attr.update({
+                slugify("{} stats".format(filt.name)): filt.stats
```

Comment:
How will the nested dict that we create for state attributes look in the frontend?

Reply:
this information is for fine tuning the filter options only. Should not be presented in the frontend.

Reply:
State attributes are shown in a table under the history graph in the more info card of sensors.

Reply:
Didn't knew that, only alternative I see is to slugify the nested dict():

outlier_erasures: 123
lowpass_median: 50

comments ?

Reply:
Probably easiest to flatten the dicts into one dict.

Reply:
yes :)